doc/source/devel/style-formatting.rst

```diff
@@ -0,0 +1,174 @@
+        If you are using the terminal, invoke ``git checkout my-new-feature`` from anywhere within the Enzo-E repository and then immediately invoke ``git pull``.
+
+.. _intermediate-level-code-formatting:
+
+At an Intermediate Level (Running Locally)
+==========================================
+
+About the precommit software
+----------------------------
+
+We rely upon the `pre-commit <https://pre-commit.com/>`__ software to actually manage these enforcement tools and checks.
+
+In more detail, `pre-commit <https://pre-commit.com/>`__ software is a framework that uses a plugin-system (configured through a file called ``.pre-commit-config.yaml``) to manage the invocation and enforcement tools.
+It was originally written to help manage when particular software is invoked by one (or more) of `git's hooks <https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks>__` .
+The software's name reflects the fact that it was primarily intended to be used with git's "pre-commit" hook.
+
+The power and utility of the pre-commit software has made it useful outside the context of git-hooks.
+In fact, we totally ignore its usage in the context of git-hooks for the remainder of the section (you are free to configure the software that way, but that choice is left up to you).
+We primarily use the pre-commit software because of its ability to automatically install locally cached copies of arbitrary versions of automated enforcement tools (this is extremely useful for tools like :ref:`clang-format <clang-format-details>`) and to easily apply all enforcement checks at once.
+
+.. note::
+
+   Because it's confusing, it's worth emphasizing that there are essentially 3 distinct entities named "pre-commit":
+
+   1. the `pre-commit <https://pre-commit.com/>`__ software.
+      Enzo-E contributors **only** need to know about this if they want to apply the enforcement tools locally.
+
+   2. the `pre-commit.ci <https://pre-commit.ci/>`__ continuous integration service.
+      This is named because the service simply executes the pre-commit software.
+      Just about all Enzo-E contributors will encounter this at one time or another.
+
+   3. the "pre-commit" `git hook <https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks>__`.
+      This is one of multiple different "hooks" offered by git.
+      The pre-commit software is named after this hook because it was originally designed to be used with this hook.
+      While developers are free to use the pre-commit software with the pre-commit hook, that is **NOT** (and is not discussed on this page)
+
+
+Running the Checks Locally
+--------------------------
+
+To run the checks locally, we encourage you need to install the pre-commit software.
+This software is written in python and can be installed with ``pip``.
+The `installation instructions <https://pre-commit.com/#installation>`__ also mention an alternative approach where you can run download and run pre-commit without fully installing it (as a "zipapp").
```

Comment:
Are there any known conflicts with the requirements for pre-commit and those of other things Enzo-E developers are likely to use, like yt?  (I don't know if that's something that has to be explicitly stated unless there's something that is likely to trip up a lot of developers.)

Reply:
I have no idea. I imagine the answer is "no" since yt also uses pre-commit